src/charm.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+
+"""Charm the calibre-web application."""
+
+import logging
+import typing
+from pathlib import Path
+from typing import cast
+
+import ops
+
+CONTAINER_NAME = 'calibre-web'
+SERVICE_NAME = 'calibre-web'
+STORAGE_NAME = 'books'
+LIBRARY_WRITE_ACTION = 'library-write'
+LIBRARY_WRITE_CONFIG = 'library-write'
+LIBRARY_INFO_ACTION = 'library-info'
+LIBRARY_INFO_FORMAT_PARAM = 'format'
+
+LibraryInfoFormat: typing.TypeAlias = typing.Literal['tree', 'ls-1']
+LIBRARY_INFO_FORMATS = typing.get_args(LibraryInfoFormat)
+
+LibraryWriteBehaviour: typing.TypeAlias = typing.Literal['skip', 'clean']
+LIBRARY_WRITE_BEHAVIOURS = typing.get_args(LibraryWriteBehaviour)
+
+
+CalibreWebLayerDict = typing.TypedDict(
+    'CalibreWebLayerDict',
+    {
+        'summary': str,
+        'description': str,
+        'services': dict[str, ops.pebble.ServiceDict],
+    },
+    total=True,
+)
+
+
+class CalibreWebCharm(ops.CharmBase):
+    """Charm the calibre-web application."""
+
+    def __init__(self, framework: ops.Framework) -> None:
+        super().__init__(framework)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        framework.observe(self.on[CONTAINER_NAME].pebble_ready, self._on_pebble_ready)
+        framework.observe(self.on[STORAGE_NAME].storage_attached, self._on_storage_attached)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on[LIBRARY_WRITE_ACTION].action, self._on_library_write)
+        framework.observe(self.on[LIBRARY_INFO_ACTION].action, self._on_library_info)
```

Comment:
It only really matters if you're going to continue with this charm after onboarding, but `action-set` does have a size limit (I don't recall what it is offhand), and I think that would potentially get hit here with a large library. I assume this is more of an example action anyway, though?

Reply:
Yeah I don't think this is a particularly useful action really, but it was fun to write. I'll look into how to handle `library-info` for larger libraries if I continue with this charm. Thanks.